staging/src/k8s.io/kubectl/pkg/cmd/auth/cani.go

```diff
@@ -309,28 +356,64 @@ func (o *CanIOptions) RunAccessCheck() (bool, error) {
 	return response.Status.Allowed, nil
 }
 
-func (o *CanIOptions) resourceFor(mapper meta.RESTMapper, resourceArg string) schema.GroupVersionResource {
-	if resourceArg == "*" {
-		return schema.GroupVersionResource{Resource: resourceArg}
+func (o *CanIOptions) resourceFor(mapper meta.RESTMapper, resourceArg string, subresourceArg string) schema.GroupVersionResource {
+	gvrNonStandardResourceSupplier := func(resource, subresource string) (schema.GroupVersionResource, error) {
+		if resource == "*" {
+			return schema.GroupVersionResource{Resource: resource}, nil
+		}
+
+		_, groupResource := schema.ParseResourceArg(strings.ToLower(resource))
+
+		for _, nonStandardResource := range nonStandardResources {
+			if (nonStandardResource.groupResource.Resource == groupResource.Resource || nonStandardResource.ResourceSingular == groupResource.Resource) &&
+				(nonStandardResource.groupResource.Group == groupResource.Group || groupResource.Group == "") {
+				var retErr error
+				if len(subresource) > 0 && !nonStandardResource.acceptsSubResource {
+					retErr = fmt.Errorf("the server doesn't have a subresource %s\n", groupResourceDescription(schema.GroupResource{
+						Group:    nonStandardResource.groupResource.Group,
+						Resource: fmt.Sprintf("%v/%v", nonStandardResource.groupResource.Resource, subresource),
+					}))
+				}
+				return schema.GroupVersionResource{Group: nonStandardResource.groupResource.Group, Resource: nonStandardResource.groupResource.Resource}, retErr
+			}
+		}
+		return schema.GroupVersionResource{}, nil
 	}
 
-	fullySpecifiedGVR, groupResource := schema.ParseResourceArg(strings.ToLower(resourceArg))
-	gvr := schema.GroupVersionResource{}
-	if fullySpecifiedGVR != nil {
-		gvr, _ = mapper.ResourceFor(*fullySpecifiedGVR)
+	gvrSupplier := func(mapper meta.RESTMapper, resource string) (schema.GroupVersionResource, error) {
+		fullySpecifiedGVR, groupResource := schema.ParseResourceArg(strings.ToLower(resource))
+
+		if fullySpecifiedGVR != nil {
+			if gvr, err := mapper.ResourceFor(*fullySpecifiedGVR); err == nil && !gvr.Empty() {
+				return gvr, nil
+			}
+		}
+		if gvr, err := mapper.ResourceFor(groupResource.WithVersion("")); err == nil && !gvr.Empty() {
+			return gvr, nil
+		}
+
+		return schema.GroupVersionResource{Group: groupResource.Group, Resource: groupResource.Resource},
+			fmt.Errorf("the server doesn't have a resource type %s\n", groupResourceDescription(groupResource))
+	}
+
+	// try to match a non-standard resource first
+	gvr, err := gvrNonStandardResourceSupplier(resourceArg, subresourceArg)
```

Comment:
I think we should try to resolve non standard resources first before we match them via a mapper. So for example we will choose `users` before e.g. `users.user.openshift.io`. This is a breaking change, but IMO it makes more sense to prefer the builtins. It certainly makes more sense for the [impersonate](https://kubernetes.io/docs/reference/access-authn-authz/authentication/#user-impersonation) verb. I suppose we could use builtins only for this verb and use custom resources for the other verbs, but I do not like the inconsistency we might introduce. If people want to check for their resources they can always use fully qualified names.

@liggitt thoughts?

Reply:
I'm actually pretty skeptical about doing all this work and surfacing way more warnings in general. The authorization system and API is intentionally opaque ... making the kubectl command "smarter" gets in the way of using the authorization system that way, and hard-coding in exceptions doesn't work well in the face of client/server skew.

Is there evidence that these additional warnings are needed, and the benefits outweigh the increased friction for people using this command to check synthetic permissions?

Reply:
I think we should first discuss the resource resolution before the we decide on the warnings. At least a couple of users think these warnings are useful, but I have no other data to back that up. The warnings are just an add-on to and do not change the authorization system.

What changes the authorization system is the resource resolution according to the the mapper/discovery. It may have different results in case user just specifies `resource` instead of `resource.version.group` or `resource.group`. I am not sure we can backtrack on this (make it less smart) and force users to always fully specify the resource, as that would break the command backwards compatibility.  In this respect, I think we have a bug as it is not possible to check for the non standard resources (`users`, and the other 3) at the moment, and this is the main reason why the PR is more complicated.

I do not like the hard coding of the values, but it seems like an okay solution at this time to me. I suppose we could include this in the discovery or alternate mechanism. Although that seems out of scope for this PR. Do you think it makes sense to propose such a feature?

Reply:
@liggitt ^